### PR TITLE
8299327: Allow super late barrier expansion of store barriers in C2

### PR DESCRIPTION
**Pull request by fisk**
ZGC uses super late barrier expansion for load barriers in C2. Generational ZGC needs to do the same thing but for store barriers.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299327](https://bugs.openjdk.org/browse/JDK-8299327): Allow super late barrier expansion of store barriers in C2

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [de45ec52](https://git.openjdk.org/jdk/pull/11779/files/de45ec526d2a5575be6438d0c52817375d4b0f89)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11779/head:pull/11779` \
`$ git checkout pull/11779`

Update a local copy of the PR: \
`$ git checkout pull/11779` \
`$ git pull https://git.openjdk.org/jdk pull/11779/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11779`

View PR using the GUI difftool: \
`$ git pr show -t 11779`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11779.diff">https://git.openjdk.org/jdk/pull/11779.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Seems fine.

**Comment by fisk** [comment]
Thanks for the review, @vnkozlov!

File: src/hotspot/cpu/ppc/ppc.ad
Comment (robcasloz) [inline src/hotspot/cpu/ppc/ppc.ad:6538]: The ADLC peephole replacement logic is not barrier data-aware as far as I can see. I wonder if barrier data is propagated correctly by peephole rules like the one in https://github.com/openjdk/jdk/blob/17a3f0e2577f2f9eb3fe62a4b8261e3dbe4c3b28/src/hotspot/cpu/ppc/ppc.ad#L14877-L14881. But I guess this is a question for PPC maintainers.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (robcasloz) [inline src/hotspot/cpu/aarch64/aarch64.ad:7976]: Why does the instruction below (`storeimmP0_volatile`) not require a similar predicate? Just checking - I guess the omission is intentional since it follows from the ZGC repo.

File: src/hotspot/share/opto/graphKit.cpp
Comment (robcasloz) [inline src/hotspot/share/opto/graphKit.cpp:1569]: Do we need this check? The analogous load construction logic (https://github.com/openjdk/jdk/blob/17a3f0e2577f2f9eb3fe62a4b8261e3dbe4c3b28/src/hotspot/share/opto/memnode.cpp#L951) does not have it.

**Review by robcasloz — APPROVED** [review]
Hi Erik, god fortsättning! Looks good, I just have a few (minor) comments. Do not forget to update the copyright headers.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (fisk) [reply src/hotspot/cpu/aarch64/aarch64.ad:7976]: No, you are right - it does need a similar predicate. The storeimmP0_volatile rule seems to be quite recent, and I wrote this code before it existed and then we must have merged without noticing. Nice catch!

File: src/hotspot/cpu/ppc/ppc.ad
Comment (fisk) [reply src/hotspot/cpu/ppc/ppc.ad:6538]: Indeed. @TheRealMDoerr would you mind having a look at this?

File: src/hotspot/share/opto/graphKit.cpp
Comment (fisk) [reply src/hotspot/share/opto/graphKit.cpp:1569]: I'll remove the check for consistency.

**Comment by fisk** [comment]
Thanks for the review, @robcasloz. I created a patch incorporating your feedback.

**Review by robcasloz — APPROVED** [review]
Thanks for addressing my comments Erik, looks good!

File: src/hotspot/cpu/ppc/ppc.ad
Comment (TheRealMDoerr) [reply src/hotspot/cpu/ppc/ppc.ad:6538]: `OptoPeephole` is a `develop_pd` flag which is only enabled on x86. So, we don't support it for PPC64.
Even if somebody switched it on for PPC64, I don't think it would break anything.
The predicates are used during the matching phase. The peephole optimizer runs later at a point of time at which the barrier data is no longer used.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (TheRealMDoerr) [reply src/hotspot/cpu/aarch64/aarch64.ad:7976]: Yeah, nice catch! This would have been a bad bug. SATB could have missed Oops which were overwritten by Store0.

File: src/hotspot/cpu/ppc/ppc.ad
Comment (fisk) [reply src/hotspot/cpu/ppc/ppc.ad:6538]: Ah, good. Thanks for checking.

**Pull request closed without merging**
